state/logs.go

```diff
@@ -74,14 +76,130 @@ func logCollectionName(modelUUID string) string {
 	return logsCPrefix + modelUUID
 }
 
-// InitDbLogs sets up the indexes for the logs collection. It should
-// be called as state is opened. It is idempotent.
-func InitDbLogs(session *mgo.Session, modelUUID string) error {
+// InitDbLogs sets up the capped collections for the logging, along with the
+// indexes for the logs collection. It should be called as state is opened. It
+// is idempotent.
+func InitDbLogs(session *mgo.Session) error {
+	// Read the capped collection size from controller config.
+	size, err := modelLogsSize(session)
+	if errors.Cause(err) == mgo.ErrNotFound {
+		// We are in early stages of database initialization, so nothing to do
+		// here. The controller model and the default model (most likely) will
+		// be initialized during bootstrap, and when the machine agent starts
+		// properly, the logs collections will be either created for them,
+		// or converted to capped collections.
+		logger.Infof("controller settings not found, early stage initialization assumed")
+		return nil
+	} else if err != nil {
+		return errors.Trace(err)
+	}
+
+	models, err := modelUUIDs(session)
+	if err != nil {
+		return err
+	}
+
+	encounteredError := false
+	for _, uuid := range models {
+		if err := InitDbLogsForModel(session, uuid, size); err != nil {
+			encounteredError = true
+			logger.Errorf("unable to initialize model logs: %v", err)
+		}
+	}
+
+	if encounteredError {
+		return errors.New("one or more errors initializing logs")
+	}
+	return nil
+}
+
+// modelLogsSize reads the model-logs-size value from the controller
+// config document and returns it. If the value isn't found the default
+// size value is returned.
+func modelLogsSize(session *mgo.Session) (int, error) {
+	// This is executed very early in the opening of the database, so there
+	// is no State, Controller, nor StatePool objects just now. Use low level
+	// mgo to access the settings.
+	var doc settingsDoc
+	err := session.DB(jujuDB).C(controllersC).FindId(controllerSettingsGlobalKey).One(&doc)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	// During initial migration there is no guarantee that the value exists
+	// in the settings document.
+	if value, ok := doc.Settings[controller.ModelLogsSize]; ok {
+		if s, ok := value.(string); ok {
+			size, _ := utils.ParseSize(s)
```

Comment:
I guess we have already validated by this time.

Reply:
Yep.